### PR TITLE
Issue #9 Mark an item as purchased

### LINKED ISSUE #9: 9. As a user, I want to mark an item on my shopping list as purchased so the app can learn how often I buy different items
## Summary
Users need a UI that allows them to mark the their items as purchased, so they can track what on their list they do and do not need to buy.

## Acceptance criteria
- [x] The `ListItem` component renders a checkbox with a semantic `<label>`.
- [x] Checking off the item in the UI also updates the `dateLastPurchased` and `totalPurchases` properties on the corresponding Firestore document
- [x] The item is shown as checked for 24 hours after the purchase is made (i.e. we assume the user does not need to buy the item again for at least 1 day). After 24 hours, the item unchecks itself so the user can buy it again.
- [x] The `updateItem` function in `firebase.js` has been filled out, and sends updates to the firestore database when an item is checked

## Notes:
- While you will need to update multiple parts of the Firestore document for this feature, **do not** worry about `dateNextPurchased` yet. That will be addressed in a future issue.
- You can use the Firestore console to test that this feature is working correctly, e.g., by manipulating the value of `dateLastPurchased`

### PR DESCRIPTION
**Pull request by amandaguan-ag**
## Description

Add checkbox to allow user to mark the their items as purchased and remain purchased for 24 hours. The checkbox will unclick itself after 24 hours.
## Related Issue

close #9 
## Acceptance Criteria
- [x] The `ListItem` component renders a checkbox with a semantic `<label>`.
- [x] Checking off the item in the UI also updates the `dateLastPurchased` and `totalPurchases` properties on the corresponding Firestore document
- [x] The item is shown as checked for 24 hours after the purchase is made (i.e. we assume the user does not need to buy the item again for at least 1 day). After 24 hours, the item unchecks itself so the user can buy it again.
- [x] The `updateItem` function in `firebase.js` has been filled out, and sends updates to the firestore database when an item is checked

## Type of Changes

enhancement

### Before

image

### After

image

## Testing Steps / QA Criteria
- From your terminal, pull down this branch with `git pull origin ag-lc-mark-as-purchase` and check out with `git checkout ag-lc-mark-as-purchase`
- Then enter `npm start`
- Click an item in List view, the dateLastPurchase should be updated with the time you click
- Go to firstore to change the dateLastPurchase to a date that is pasted within 24 hours (eg feb 28th 1pm -> feb 27th 10pm) the dateLastPurchase is , the item should be remained clicked
- Go to firstore to change the dateLastPurchase to a date that past over 24 hours(eg feb 28th 1pm -> feb 1st 1pm) , the item should be automatically unclicked

### CODE REVIEW

**Review by g-lee2 — COMMENTED** [review]
Hey! Nice work! So the site breaks when there are items in a shopping list that are missing a dateLastPurchased because of the dateLastPurchased.seconds dependency in the useEffect. Using dateLastPurchased instead fixes this. Also, considering the 24-hour checkbox disable logic, the else statement in handleChange seems redundant.

File: src/components/ListItem.jsx
Comment (MiliMade) [inline src/components/ListItem.jsx:1]: Just echoing a what @g-lee2 said earlier about the useEffect dependency on line 35 - I'm getting a blank screen in the list view. Once I changed ```dateLastPurchased.seconds``` to just ```dateLastPurchased```, everything seemed to work. I was able to see the list items and the empty checkboxes next to them.  Clicking the box gives me a prompt to confirm purchase and the box is then checked and greyed out.

**Review by MiliMade — CHANGES_REQUESTED** [review]
Might need to change the dependency on useEffect

File: src/components/ListItem.jsx
Comment (chungleee) [reply src/components/ListItem.jsx:1]: nice catch! 

edited to `dateLastPurchased` and removed the `else` in `handleChange`

**Review by g-lee2 — APPROVED** [review]

**Review by MiliMade — APPROVED** [review]

**Review by MiliMade — APPROVED** [review]

File: src/components/ListItem.jsx [hunk 1]
```diff
@@ -1,5 +1,62 @@
...
+import { updateItem } from '../api/firebase';
+import { ONE_DAY_IN_MILLISECONDS } from '../utils/dates';
 
-export function ListItem({ name }) {
-	return <li className="ListItem">{name}</li>;
+export function ListItem({ name, listPath, itemId, dateLastPurchased }) {
+	const [isChecked, setIsChecked] = useState(false);
+	const [expired, setExpired] = useState();
+
+	const handleChange = async () => {
```
Comment (djtaylor8) [inline src/components/ListItem.jsx:10]: Nice implementation here! I know this isn't explicitly in the AC, but I think this provides a better user experience to make sure a user doesn't accidentally check something.

File: src/components/ListItem.jsx [hunk 2]
```diff
@@ -1,5 +1,62 @@
...
+			if (isChecked) {
+				await updateItem(listPath, itemId);
+			}
+		};
+		try {
+			handleUpdateItem();
+		} catch (error) {
+			console.log('error: ', error);
+		}
+	}, [isChecked]);
```
Comment (djtaylor8) [inline src/components/ListItem.jsx]: I'm seeing a few missing dependencies from this array:

Screenshot 2024-03-01 at 3 04 39 PM

**Review by djtaylor8 — APPROVED** [review]
Really nice job on this! Left a minor comment about adding missing items to a dependency array, but otherwise this is looking good. 

I would also recommend adding in a few more comments to your code to provide some more clarity. You don't need to add much, but they can help clarify things in the future and also it's good documentation practice.

File: src/api/firebase.js [hunk 3]
```diff
@@ -178,7 +179,13 @@ export async function addItem(listPath, { itemName, daysUntilNextPurchase }) {
...
-export async function updateItem() {
+export async function updateItem(listPath, itemId) {
+	const listCollectionRef = doc(db, `${listPath}/items`, itemId);
+
+	await updateDoc(listCollectionRef, {
+		dateLastPurchased: new Date(),
+		totalPurchases: increment(1),
+	});
 	/**
 	 * TODO: Fill this out so that it uses the correct Firestore function
```
Comment (lyleschemmerling) [inline src/api/firebase.js]: We should take out the TODO now that we filled in the function definition

**Review by lyleschemmerling — APPROVED** [review]
Tested and looks like it's work working great. Good job!

**Review by RachaelSkye — APPROVED** [review]
Great work!

**Pull request merged**